[js] Enable js.Browser to work with WebWorkers

haxiomic opened the pull request:
JS WebWorkers do not have access to the same global scope as regular JS execution

**Changes**

- Add `self` to js.Browser. In browser js there's a global [`self`](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/self) object, which links to the global scope (like `window`). In a WebWorker, the `window` field is not available, instead `self` is used to access the global scope (and access worker-specific fields)
- Replace usage of `window` with `js.Lib.global` so that objects like console and navigator are accessible from a worker context

closes #4435, #9168

haxiomic committed:
[js] add js.Browser.self, closes #4435

haxiomic committed:
change self type to common root: WorkerGlobalScope

RealyUniqueName commented on std/js/Browser.hx:
```diff
@@ -26,6 +26,12 @@ import js.html.Storage;
+	/** The global scope with additional fields available only in a worker context. */
+	public static var self(get, never):js.html.WorkerGlobalScope;
+
+	extern inline static function get_self()
```
Add a type hint for return type. 
Otherwise it is typed as a monomorph (because of `untyped` return) and could possibly be bound to whatever first type it gets unified with.

haxiomic committed:
[js] add type-hint for get_self

Simn commented on std/js/Browser.hx:
```diff
@@ -26,6 +26,12 @@ import js.html.Storage;
+	public static var self(get, never):js.html.WorkerGlobalScope;
+
+	extern inline static function get_self():js.html.WorkerGlobalScope
+		return untyped __js__("self");
```
Please no more `untyped __js__` in new additions. There's `js.Sytnax.code` and such to deal with this.

RealyUniqueName commented:
After revisiting this PR I find this approach unreliable.
If there's a local var named `self`, then `Browser.self` would return that var instead of a proper `self` object.

haxiomic commented:
@RealyUniqueName that’s a good point, it’s currently using the same approach for Browser.window
Although I can’t figure out a way to avoid the possible local variable collision - do you have any thoughts on an alternative approach?

Maybe rather than a getter we assign to `self` at startup, that way it’s namespaced to js.Browser.self

Aurel300 commented:
Can we not add `window` and `self` to JS keywords so that locals are never named that?

RealyUniqueName commented:
Using non-inlinable getters would solve the problem with a local var shadowing.
It still could conflict with a global var defined somewhere in userland js, but that's a neglible probability I think.

Aurel300 commented:
You can't declare a variable called `window` (at least Firefox doesn't let me):

```js
>>> let window = "3"
SyntaxError: redeclaration of non-configurable global property window
```

I assume in worker scopes this is the same for `self`?

RealyUniqueName commented:
`self` is allowed to be declared in non-worker context. That's enough for `Browser.self` to return wrong references.

RealyUniqueName commented:
As @Aurel300 said, we could add `self` to js keywords list.
`window` is already in the list.

haxiomic commented:
Just another discovery to add to this story. [`globalThis`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis) is intended to be a standardised way to access the global-scope that also works in node.js (where self and window don't work). 

The kicker: it's not 100% supported right now https://caniuse.com/#search=globalThis

RealyUniqueName commented:
We already have `js.Lib.global`:
https://github.com/HaxeFoundation/haxe/blob/c79046aee5ba2f18ee9ff868fd225e106fa02f32/std/js/Lib.hx#L114-L120

haxiomic commented:
@RealyUniqueName did not know about that one, that's perfect and solves the issue @filt3rek was having trying to access worker-scope :)

I think having a Browser.self is useful nevertheless because it gives the correct types for the worker-context global scope so I'll finish off this PR

haxiomic committed:
[js] Switch from untyped to js.Syntax.code

haxiomic committed:
Use js.Lib.global in place of self to avoid possible local var conflict

haxiomic commented:
Updated this one:
- to avoid the possible local var conflict I've switched to `js.Lib.global`
- Replaced use of `window` with `js.Lib.global` where the field is also available in in a worker context (e.g. console)

haxiomic committed:
Replace usage of window with js.Lib.global where API is available in a worker context

haxiomic committed:
Correct comment

haxiomic committed:
Merge branch 'development' into js-browser-self

haxiomic commented:
I think this one is good to go now, using haxe's $global solves the issue we were having
